File: DragonIsland-GP2X/SoundEffect.cpp
#include "SoundEffect.h"

SoundEffect::SoundEffect()
{
	   mSound = new Mix_Chunk[16];
	   mSound[0] = *Mix_LoadWAV("mnt/sd/Dragon Island/sfx/block.wav");
	   mSound[1] = *Mix_LoadWAV("mnt/sd/Dragon Island/sfx/coin.wav");
	   mSound[2] = *Mix_LoadWAV("mnt/sd/Dragon Island/sfx/die.wav");
	   mSound[3] = *Mix_LoadWAV("mnt/sd/Dragon Island/sfx/fire.wav");
	   mSound[4] = *Mix_LoadWAV("mnt/sd/Dragon Island/sfx/hit.wav");
	   mSound[5] = *Mix_LoadWAV("mnt/sd/Dragon Island/sfx/jump.wav");
	   mSound[6] = *Mix_LoadWAV("mnt/sd/Dragon Island/sfx/kill.wav");
	   mSound[7] = *Mix_LoadWAV("mnt/sd/Dragon Island/sfx/option.wav");
	   mSound[8] = *Mix_LoadWAV("mnt/sd/Dragon Island/sfx/powerdown.wav");
	   mSound[9] = *Mix_LoadWAV("mnt/sd/Dragon Island/sfx/powerup.wav");
	   mSound[10] = *Mix_LoadWAV("mnt/sd/Dragon Island/sfx/select.wav");
	   mSound[11] = *Mix_LoadWAV("mnt/sd/Dragon Island/sfx/speech.wav");
	   mSound[12] = *Mix_LoadWAV("mnt/sd/Dragon Island/sfx/tank.wav");
}

void SoundEffect::play(string filename, int repeat)
{
	int channel;
	if (Settings::Sound) {
		if (filename == "block.wav")
			channel = Mix_PlayChannel(-1, &mSound[0], repeat);
		else if (filename == "coin.wav")
			channel = Mix_PlayChannel(-1, &mSound[1], repeat);
		else if (filename == "die.wav")
			channel = Mix_PlayChannel(-1, &mSound[2], repeat);
		else if (filename == "fire.wav")
			channel = Mix_PlayChannel(-1, &mSound[3], repeat);
		else if (filename == "hit.wav")
			channel = Mix_PlayChannel(-1, &mSound[4], repeat);
		else if (filename == "jump.wav")
			channel = Mix_PlayChannel(-1, &mSound[5], repeat);
		else if (filename == "kill.wav")
			channel = Mix_PlayChannel(-1, &mSound[6], repeat);
		else if (filename == "option.wav")
			channel = Mix_PlayChannel(-1, &mSound[7], repeat);
		else if (filename == "powerdown.wav")
			channel = Mix_PlayChannel(-1, &mSound[8], repeat);
		else if (filename == "powerup.wav")
			channel = Mix_PlayChannel(-1, &mSound[9], repeat);
		else if (filename == "select.wav")
			channel = Mix_PlayChannel(-1, &mSound[10], repeat);
		else if (filename == "speech.wav")
			channel = Mix_PlayChannel(-1, &mSound[11], repeat);
		else if (filename == "tank.wav")
			channel = Mix_PlayChannel(-1, &mSound[12], repeat);		

		if(channel == -1) {
			fprintf(stderr, "Unable to play WAV file: %s\n", Mix_GetError());
		}
	}

}

File: DragonIsland-Android-Studio/app/src/main/java/com/citex/dragonisland/android/audio/SoundEffect.java
package com.citex.dragonisland.android.audio;

import java.io.IOException;
import java.util.HashMap;

import android.annotation.SuppressLint;
import android.content.res.AssetFileDescriptor;
import android.media.AudioManager;
import android.media.SoundPool;

import com.citex.dragonisland.android.MainActivity;
import com.citex.dragonisland.core.audio.ISoundEffect;
import com.citex.dragonisland.core.game.Settings;

/**
 * SoundEffect.java
 * This class plays wav sounds effects.
 * Copyright (C) 2023 Lawrence Schmid
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

public class SoundEffect implements ISoundEffect {

	/** Sound pool. */
	private static SoundPool mSoundPool;
	
	/** Sound pool map. */
	private static HashMap<Integer, Integer> mSoundPoolMap;
	
	/**
	 * Initialise a collection of sound effects.
	 * @param path Path to the wav sound effect files.
	 * @throws IOException 
	 * @throws IllegalStateException 
	 */
	@SuppressLint("UseSparseArrays") 
	public SoundEffect(String path) throws IllegalStateException, IOException {

		// Initialise the sound pool.
		mSoundPool = new SoundPool(4, AudioManager.STREAM_MUSIC, 100);
		
		// Sound pool map.
		mSoundPoolMap = new HashMap<>(12);
		
		// Load sound effects from assets.
		AssetFileDescriptor afd = MainActivity.mAssetManager.openFd(path + "block.wav");
		mSoundPoolMap.put(1, mSoundPool.load(afd, 1));
		
		afd = MainActivity.mAssetManager.openFd(path + "coin.wav");
		mSoundPoolMap.put(2, mSoundPool.load(afd, 1));
		
		afd = MainActivity.mAssetManager.openFd(path + "die.wav");
		mSoundPoolMap.put(3, mSoundPool.load(afd, 1));
		
		afd = MainActivity.mAssetManager.openFd(path + "fire.wav");
		mSoundPoolMap.put(4, mSoundPool.load(afd, 1));
		
		afd = MainActivity.mAssetManager.openFd(path + "hit.wav");
		mSoundPoolMap.put(5, mSoundPool.load(afd, 1));
		
		afd = MainActivity.mAssetManager.openFd(path + "jump.wav");
		mSoundPoolMap.put(6, mSoundPool.load(afd, 1));
		
		afd = MainActivity.mAssetManager.openFd(path + "kill.wav");
		mSoundPoolMap.put(7, mSoundPool.load(afd, 1));
		
		afd = MainActivity.mAssetManager.openFd(path + "option.wav");
		mSoundPoolMap.put(8, mSoundPool.load(afd, 1));
		
		afd = MainActivity.mAssetManager.openFd(path + "powerdown.wav");
		mSoundPoolMap.put(9, mSoundPool.load(afd, 1));
		
		afd = MainActivity.mAssetManager.openFd(path + "powerup.wav");
		mSoundPoolMap.put(10, mSoundPool.load(afd, 1));
		
		afd = MainActivity.mAssetManager.openFd(path + "select.wav");
		mSoundPoolMap.put(11, mSoundPool.load(afd, 1));

	}

	/**
	 * Plays a wav sound effect 
	 * @param soundName File name of the sound effect.
	 */
	public void play(String soundName) {
		
		// Initialise sound properties.
		float leftVolume = 0.5f;
		float rightVolume = 0.5f;
		int priority = 1;
		int loop = 0;
		float rate = 1.0f;
		
		if (Settings.Sound) {

			switch (soundName) {
				case "block.wav":
					mSoundPool.play(mSoundPoolMap.get(1), leftVolume, rightVolume, priority, loop, rate);
					break;
				case "coin.wav":
					mSoundPool.play(mSoundPoolMap.get(2), leftVolume - 0.05f, rightVolume - 0.05f, priority, loop, rate);
					break;
				case "die.wav":
					mSoundPool.play(mSoundPoolMap.get(3), leftVolume - 0.05f, rightVolume - 0.05f, priority, loop, rate);
					break;
				case "fire.wav":
					mSoundPool.play(mSoundPoolMap.get(4), leftVolume, rightVolume, priority, loop, rate);
					break;
				case "hit.wav":
					mSoundPool.play(mSoundPoolMap.get(5), leftVolume, rightVolume, priority, loop, rate);
					break;
				case "jump.wav":
					mSoundPool.play(mSoundPoolMap.get(6), leftVolume - 0.1f, rightVolume - 0.1f, priority, loop, rate);
					break;
				case "kill.wav":
					mSoundPool.play(mSoundPoolMap.get(7), leftVolume, rightVolume, priority, loop, rate);
					break;
				case "option.wav":
					mSoundPool.play(mSoundPoolMap.get(8), leftVolume, rightVolume, priority, loop, rate);
					break;
				case "powerdown.wav":
					mSoundPool.play(mSoundPoolMap.get(9), leftVolume, rightVolume, priority, loop, rate);
					break;
				case "powerup.wav":
					mSoundPool.play(mSoundPoolMap.get(10), leftVolume, rightVolume, priority, loop, rate);
					break;
				case "select.wav":
					mSoundPool.play(mSoundPoolMap.get(11), leftVolume, rightVolume, priority, loop, rate);
					break;
			}

		}
	}

    /**
     * Destroys the resources.
     */
	public void destroy() {
		
		// Destroy sound effects.
		mSoundPool.release();
		
	}
}
